client/app/bundles/course/assessment/submission/components/ScribingView/ScribingCanvas.jsx

```diff
@@ -0,0 +1,549 @@
+    }
+  }
+
+  // Helpers
+
+  // Function Helpers
+
+  disableObjectSelection() {
+    this.canvas.forEachObject(object => (
+      object.selectable = false // eslint-disable-line no-param-reassign
+    ));
+  }
+
+  // This method only enable selection for interactive texts
+  // enableTextSelection() {
+  //   this.canvas.clear();
+  //   this.initializeScribbles(false);
+  //   this.canvas.forEachObject(object => (
+  //     // eslint-disable-next-line no-param-reassign
+  //     object.selectable = (object.type === 'i-text')
+  //   ));
+  // }
+
+  // This method clears the selection-disabled scribbles
+  // and reloads them to enable selection again
+  enableObjectSelection() {
+    this.canvas.clear();
+    this.initializeScribbles(false);
+  }
+
+  setSelectedShape = (shape) => {
+    this.setState({
+      selectedShape: shape,
+    });
+  }
+
+  // Legacy code needed to support migrated v1 scribing questions.
+  // This code scales/unscales the scribbles by a standard number.
+  normaliseScribble(scribble, isDenormalise) {
+    const STANDARD = 1000;
+    let factor;
+
+    if (isDenormalise) {
+      factor = this.canvas.getWidth() / STANDARD;
+    } else {
+      factor = STANDARD / this.canvas.getWidth();
+    }
+
+    scribble.set({
+      scaleX: scribble.scaleX * factor,
+      scaleY: scribble.scaleY * factor,
+      left: scribble.left * factor,
+      top: scribble.top * factor,
+    });
+  }
+
+  denormaliseScribble(scribble) {
+    return this.normaliseScribble(scribble, true);
+  }
+
+  initializeScribbles = (isFirstInit) => {
+    const { scribbles } = this.props.scribing.answer;
+    const userId = this.props.scribing.answer.user_id;
+
+    this.isScribblesLoaded = false;
+    let userScribble = [];
+
+    if (isFirstInit) {
+      this.canvas.layers = [];
+    } else {
+      this.canvas.layers.forEach(layer => this.canvas.add(layer.scribbleGroup));
+    }
+
+    if (scribbles) {
+      scribbles.forEach((scribble) => {
+        const objects = JSON.parse(scribble.content).objects;
+        const fabricObjs = [];
+
+        // Parse JSON to Fabric.js objects
+        for (let i = 0; i < objects.length; i++) {
+          // eslint-disable-next-line no-undef
+          const klass = fabric.util.getKlass(objects[i].type);
+          switch (objects[i].type) {
+            case 'path': {
+              klass.fromObject(objects[i], (obj) => {
+                this.denormaliseScribble(obj);
+                fabricObjs.push(obj);
+              });
+              break;
+            }
+            case 'i-text':
+            case 'line':
+            case 'rect':
+            case 'ellipse': {
+              const obj = klass.fromObject(objects[i]);
+              this.denormaliseScribble(obj);
+              fabricObjs.push(obj);
+              break;
+            }
+            default: {
+              break;
+            }
+          }
+        }
+
+        // Create layer for each user's scribble
```

Comment:
Seems a layer is created for every line/rect/path etc, not each user. This causes a 'non-unique component key' error when the 'layers' button in clicked.